### PR TITLE
Make ldmsd continue startup on producer's hostname resolution failure

### LINKED ISSUE #1489: ldmsd aggregator dies when a hostname of a sampler node can't be resolved
ldmsd aggregator dies when a hostname of a sampler node can't be resolved 

```
ldmsd_prdcr_new: my_bad_hostname:6002 not resolved.
msg_no 4654: error 97: Error resolving hostname 'my_bad_hostname:'
```

I would like for ldmsd to continue working and simply kick the bad host out of the list.

Version: ovis-4.4.2

### PR DESCRIPTION
**Pull request by nichamon**
Change producer's hostname resolution failure at config time from an error to a warning, allowing ldmsd to continue starting up. The hostname will be resolved again before establishing the connection and retired at the reconnect interval if the resolution fails.

Resolves #1489

### CODE REVIEW

File: ldms/src/ldmsd/ldmsd_prdcr.c [hunk 1]
```diff
@@ -845,6 +845,17 @@ static void prdcr_connect(ldmsd_prdcr_t prdcr)
 {
 	int ret;
 
+	if (prdcr_resolve(prdcr->host_name, prdcr->port_no, &prdcr->ss, &prdcr->ss_len)) {
+		if (0 == prdcr->not_resolved_cnt) {
+			ldmsd_log(LDMSD_LERROR, "Producer '%s' connection failed. Unable to resolve hostname '%s:%u'.\n",
+				prdcr->obj.name, prdcr->host_name,(unsigned) prdcr->port_no);
+		}
+		prdcr->not_resolved_cnt++;
```
Comment (tom95858) [inline ldms/src/ldmsd/ldmsd_prdcr.c:853]: @nichamon, We should talk about this. I don't agree that we should only report this once. The service comes and goes and we only have one user who complained about the possibility of their configuration error being reported over and over. Let's consult with other users in the UG meeting on Monday.

**Comment by tom95858** [comment]
@jstile-lbl, this is the change that modifies the handling of unresolved host names causing the daemon to exit. Please weigh in on wether the error should be reported just once or every reconnect interval. There will be a reconnect error anyway, it seems like the administrator should know why the reconnect failed.

**Comment by johnstile** [comment]
By the time the administrator sees the logs, the original name resolution error may have rotated out, so I think the name resolution error should be present in each reconnect.

Additionally one could identify when the name resolution is fixed by filtering for these errors.

**Comment by nichamon** [comment]
@nichamon will make the following changes to the patch:
- [ ] Make IP cache configurable. Caching IP is the default. This is to support cloud environment that the IPs of compute nodes change when compute nodes are up.
- [ ] If caching IP is enabled, ldmsd will not resolve sampler's hostname again after it is resolved
- [ ] Make ldmsd report error messages at every reconnect until the hostname can be resolved

**Comment by nichamon** [comment]
The revision also addresses the problem that `gethostbyname()` is not thread-safe by using `getaddrinfo()`. There is one single thread that handles configuration files. However, there may be multiple threads that connect to sampler daemons.

**Comment by nichamon** [comment]
@narategithub The patch touched the failover paths. Please help me make sure that I didn't miss anything.

File: ldms/src/ldmsd/ldmsd_failover.c [hunk 2]
```diff
@@ -2079,7 +2086,7 @@ int failover_cfgprdcr_handler(ldmsd_req_ctxt_t req)
 	}
 
 	p = ldmsd_prdcr_new_with_auth(name, xprt, host, atoi(port), ptype,
-			atoi(interval), auth, _uid, _gid, _perm);
+			atoi(interval), auth, _uid, _gid, _perm, atoi(cache_ip));
```
Comment (narategithub) [inline ldms/src/ldmsd/ldmsd_failover.c:2089]: what happened if `cache_ip` is `NULL`?

Comment (baallan) [reply ldms/src/ldmsd/ldmsd_failover.c:2089]: more significantly, afaict ip may be 64 bits, so atoi is not necessarily a valid  conversion.

Comment (nichamon) [reply ldms/src/ldmsd/ldmsd_failover.c:2089]: I added the `cache_ip` value to the producer information message sent by the buddy ldmsd, so it isn't NULL, the same as the other attributes.

Comment (tom95858) [inline ldms/src/ldmsd/ldmsd_failover.c:2089]: @nichamon, did we check what would happen if we run a 4.4.5 ldmsd_controller/maestro against a 4.4.3 ldmsd? As long as it ignores the extra parameter I think we're fine, but it shouldn't crash.

**Pull request merged**